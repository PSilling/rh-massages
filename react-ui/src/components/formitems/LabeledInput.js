// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Col, FormGroup, Label, Input, Tooltip } from "reactstrap";

// util imports
import Util from "../../util/Util";

/**
 * A combined formgroup component with a label (can have a tooltip) and a input.
 * A column is included to set input size.
 */
class LabeledInput extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter") {
      this.props.onEnterPress();
    }
  };

  render() {
    const { onChange, label, value, onEnterPress, options, size, tooltip, ...rest } = this.props;
    return (
      <Col md={this.props.size}>
        <FormGroup>
          <Label id={this.tooltipTarget} for={`${this.tooltipTarget}_input`}>
            {this.props.label}
          </Label>
          <Input
            {...rest}
            id={`${this.tooltipTarget}_input`}
            value={this.props.value}
            onChange={this.props.onChange}
            onKeyPress={this.handleKeyPress}
            placeholder={this.props.label}
            list={`${this.tooltipTargets}_list`}
          />
          {this.props.options.length > 0 && (
            <datalist id={`${this.tooltipTargets}_list`}>
              {this.props.options.map(item => (
                <option key={item} value={item} />
              ))}
            </datalist>
          )}
          {this.props.tooltip !== "" && (
            <Tooltip isOpen={this.state.tooltipActive} target={this.tooltipTarget} toggle={this.toggleTooltip}>
              {this.props.tooltip}
            </Tooltip>
          )}
        </FormGroup>
      </Col>
    );
  }
}

LabeledInput.propTypes = {
  /** function called on input value change */
  onChange: PropTypes.func.isRequired,
  /** input label text */
  label: PropTypes.string.isRequired,
  /** current value of the input field */
  value: PropTypes.string.isRequired,
  /** input datalist options */
  onEnterPress: PropTypes.func,
  /** input datalist options */
  options: PropTypes.arrayOf(PropTypes.string),
  /** input column length */
  size: PropTypes.string,
  /** input label tooltip */
  tooltip: PropTypes.string
};

LabeledInput.defaultProps = {
  onEnterPress() {},
  options: [],
  size: "12",
  tooltip: ""
};

export default LabeledInput;
