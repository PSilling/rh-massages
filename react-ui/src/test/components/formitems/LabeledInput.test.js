// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import { Col, Input, Label, Tooltip } from "reactstrap";
import LabeledInput from "../../../components/formitems/LabeledInput";

test("renders content with correct props", () => {
  const testFunction = jest.fn();
  const wrapper = shallow(<LabeledInput onChange={testFunction} label="test" value="value" tooltip="tooltip" />);
  const col = wrapper.find(Col);
  const label = wrapper.find(Label);
  const input = wrapper.find(Input);
  const tooltip = wrapper.find(Tooltip);

  expect(testFunction).not.toHaveBeenCalled();

  input.props().onChange();

  expect(testFunction).toHaveBeenCalledTimes(1);

  expect(col.props().md).toEqual("12");
  expect(label.props().id).toEqual(tooltip.props().target);
  expect(label.props().for).toEqual(input.props().id);
  expect(label.props().children).toEqual("test");
  expect(input.props().value).toEqual("value");
  expect(tooltip.props().isOpen).toBe(false);
  expect(tooltip.props().children).toEqual("tooltip");
  expect(wrapper).toMatchSnapshot();
});
