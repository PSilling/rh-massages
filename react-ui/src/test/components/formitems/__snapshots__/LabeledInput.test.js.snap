// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders content with correct props 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LabeledInput
    label="test"
    onChange={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEnterPress={[Function]}
    options={Array []}
    size="12"
    tooltip="tooltip"
    value="value"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <FormGroup
        tag="div"
      >
        <Label
          for="Tooltip1_input"
          id="Tooltip1"
          tag="label"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          test
        </Label>
        <Input
          id="Tooltip1_input"
          list="undefined_list"
          onChange={
            [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            }
          }
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="test"
          type="text"
          value="value"
        />
        <Tooltip
          autohide={true}
          delay={
            Object {
              "hide": 250,
              "show": 0,
            }
          }
          hideArrow={false}
          isOpen={false}
          placement="top"
          placementPrefix="bs-tooltip"
          target="Tooltip1"
          toggle={[Function]}
        >
          tooltip
        </Tooltip>
      </FormGroup>,
      "md": "12",
      "tag": "div",
      "widths": Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Label
            for="Tooltip1_input"
            id="Tooltip1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            test
          </Label>,
          <Input
            id="Tooltip1_input"
            list="undefined_list"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="test"
            type="text"
            value="value"
          />,
          false,
          <Tooltip
            autohide={true}
            delay={
              Object {
                "hide": 250,
                "show": 0,
              }
            }
            hideArrow={false}
            isOpen={false}
            placement="top"
            placementPrefix="bs-tooltip"
            target="Tooltip1"
            toggle={[Function]}
          >
            tooltip
          </Tooltip>,
        ],
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "test",
            "for": "Tooltip1_input",
            "id": "Tooltip1",
            "tag": "label",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": "test",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "id": "Tooltip1_input",
            "list": "undefined_list",
            "onChange": [MockFunction] {
              "calls": Array [
                Array [],
              ],
              "results": Array [
                Object {
                  "isThrow": false,
                  "value": undefined,
                },
              ],
            },
            "onFocus": [Function],
            "onKeyPress": [Function],
            "placeholder": "test",
            "type": "text",
            "value": "value",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autohide": true,
            "children": "tooltip",
            "delay": Object {
              "hide": 250,
              "show": 0,
            },
            "hideArrow": false,
            "isOpen": false,
            "placement": "top",
            "placementPrefix": "bs-tooltip",
            "target": "Tooltip1",
            "toggle": [Function],
          },
          "ref": null,
          "rendered": "tooltip",
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <FormGroup
          tag="div"
        >
          <Label
            for="Tooltip1_input"
            id="Tooltip1"
            tag="label"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            test
          </Label>
          <Input
            id="Tooltip1_input"
            list="undefined_list"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            onFocus={[Function]}
            onKeyPress={[Function]}
            placeholder="test"
            type="text"
            value="value"
          />
          <Tooltip
            autohide={true}
            delay={
              Object {
                "hide": 250,
                "show": 0,
              }
            }
            hideArrow={false}
            isOpen={false}
            placement="top"
            placementPrefix="bs-tooltip"
            target="Tooltip1"
            toggle={[Function]}
          >
            tooltip
          </Tooltip>
        </FormGroup>,
        "md": "12",
        "tag": "div",
        "widths": Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Label
              for="Tooltip1_input"
              id="Tooltip1"
              tag="label"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
            >
              test
            </Label>,
            <Input
              id="Tooltip1_input"
              list="undefined_list"
              onChange={
                [MockFunction] {
                  "calls": Array [
                    Array [],
                  ],
                  "results": Array [
                    Object {
                      "isThrow": false,
                      "value": undefined,
                    },
                  ],
                }
              }
              onFocus={[Function]}
              onKeyPress={[Function]}
              placeholder="test"
              type="text"
              value="value"
            />,
            false,
            <Tooltip
              autohide={true}
              delay={
                Object {
                  "hide": 250,
                  "show": 0,
                }
              }
              hideArrow={false}
              isOpen={false}
              placement="top"
              placementPrefix="bs-tooltip"
              target="Tooltip1"
              toggle={[Function]}
            >
              tooltip
            </Tooltip>,
          ],
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "test",
              "for": "Tooltip1_input",
              "id": "Tooltip1",
              "tag": "label",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": "test",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "id": "Tooltip1_input",
              "list": "undefined_list",
              "onChange": [MockFunction] {
                "calls": Array [
                  Array [],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              },
              "onFocus": [Function],
              "onKeyPress": [Function],
              "placeholder": "test",
              "type": "text",
              "value": "value",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autohide": true,
              "children": "tooltip",
              "delay": Object {
                "hide": 250,
                "show": 0,
              },
              "hideArrow": false,
              "isOpen": false,
              "placement": "top",
              "placementPrefix": "bs-tooltip",
              "target": "Tooltip1",
              "toggle": [Function],
            },
            "ref": null,
            "rendered": "tooltip",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
